docs/Add-Models.md

```diff
@@ -0,0 +1,270 @@
+
+```python
+from transformers import PretrainedConfig as HfPretrainedConfig
+
+# ...
+
+@classmethod
+def from_hf_config(cls, hf_config: HfPretrainedConfig):
+    return LlamaConfig(
+        seq_len=hf_config.max_position_embeddings,
+        hidden_dim=hf_config.hidden_size,
+        intermediate_dim=hf_config.intermediate_size,
+        num_layers=hf_config.num_hidden_layers,
+        num_heads=hf_config.num_attention_heads,
+        activation_function=hf_config.hidden_act,
+        initializer_range=hf_config.initializer_range,
+        layer_norm_epsilon=hf_config.rms_norm_eps,
+        rope_scaling=hf_config.rope_scaling,
+    )
+```
+
+Lastly, you should register your head model's class name as a class property. This step can be deferred until the head class is constructed.
+This class property would make it easier to call your model with the config class. For example:
+
+```python
+@property
+def model_type(cls) -> Type["LlamaLMHeadModel"]:
+    return LlamaLMHeadModel
+```
+
+### Write Model
+After you have defined your config class, you can start writing your model.
+You can follow the breakdown of the model in your Hugging Face implmentation. This would make it easier to validate the implementation through unit tests.
+
+For example, in GPT2, we have the following breakdown:
+- `Gpt2Mlp`
+- `Gpt2Attention`
+- `Gpt2Block`: a block of Gpt2Mlp and Gpt2Attention
+- `Gpt2Transformer`: a stack of Gpt2Block
+- `Gpt2Embeddings`: token and position embeddings
+- `Gpt2LMHead`: a complete GPT2 model with embedding, transformer, and LM head.
+
+We follow the same breakdown in the implementation of Llama in Levanter.
+
+#### Note on the Implementation Format
+- Each class will have its key layers and components defined as attributes and be initialized with a static method `init()`.
+- Each class will inherit from `StateDictSerializationMixin` from `torch_serialization` and Equinox's `Module` class.
+
+### Mapping weights from/to Hugging Face
+To load weights from Hugging Face, you will need to write a class function `from_hf_state_dict()` in each of your model class. It takes in a Hugging Face state_dict and returns a Levanter state_dict.
+
+For implementation, there are a few helper classes from `torch_serialization` that you can use:
+- To add specific prefix to the keys of Hugging Face state_dict, you can use the helper function `apply_prefix()`. The prefix comes from the name of attributes defined at the beginning of your model class.
+- To unflatten the linear layers of Hugging Face, you can use the helper function `unflatten_linear_params()`.
+- To unstack the transformer blocks of Hugging Face, you can use the helper function `unstack_transformer_blocks()`.
+
+For example, below is the implementation of `from_hf_state_dict()` in LlamaAttention:
```

Comment:
```suggestion
For example, below is the implementation of `from_state_dict()` in `LlamaAttention`. `LLamaAttention`, like most attention layers in Levanter projects the embeddings (with shape `(Pos, Embed)`) to a tensor of shape `(Pos, Head, HeadSize)` rather than the convention in Hugging Face Transformers to use `(Pos, Head * HeadSize)` and then reshaping (because PyTorch Linear doesn't support multiple input or output dimensions).

This difference in convention means that we have to flatten linear layers when converting to a Hugging Face Transformers-compatible state dict and unflatten them when they are read in.
```